spec/expression.dd

```diff
@@ -63,19 +63,70 @@ $(P Note: Rvalues include all literals, special value keywords such as `__FILE__
 
-$(P Binary expressions are evaluated in strictly left-to-right order.
-Function arguments are evaluated in strictly left-to-right order for
-functions with `extern (D)` $(DDSUBLINK spec/attribute, linkage, linkage).)
+$(P Built-in prefix unary expressions `++` and `--` are evaluated as if lowered (rewritten) to
+assignments as follows: `++expr` becomes `((expr) += 1)`, and `--expr` becomes `((expr) -= 1)`.
+Therefore, the result of prefix `++` and `--` is the lvalue after the side effect has been
+effected.)
 
-        -------------
-        import std.conv;
-        int i = 0;
-        (i = 2) = ++i * i++ + i;
-        assert(i == 13); // left to right evaluation of side effects
-        assert(text(++i, ++i) == "1415"); // left to right evaluation of arguments
-        -------------
+$(P Built-in postfix unary expressions `++` and `--` are evaluated as if lowered (rewritten) to lambda
+invocations as follows: `expr++` becomes `(ref T x){auto t = x; ++x; return t;}(expr)`, and
+`expr--` becomes `(ref T x){auto t = x; --x; return t;}(expr)`. Therefore, the result of postfix
+`++` and `--` is an rvalue just before the side effect has been effected.)
+
+$(P Binary expressions except for $(GLINK AssignExpression), $(GLINK OrOrExpression), and
+$(GLINK AndAndExpression) are evaluated in lexical order (left-to-right). Example:)
+
+$(SPEC_RUNNABLE_EXAMPLE_RUN
+-------------
+int i = 2;
+i = ++i * i++ + i;
+assert(i == 3 * 3 + 4);
+-------------
+)
+
+$(P $(GLINK OrOrExpression) and $(GLINK AndAndExpression) evaluate their left-hand side argument
+first. Then, $(GLINK OrOrExpression) evaluates its right-hand side if and only if its left-hand
+side does not evaluate to nonzero. $(GLINK AndAndExpression) evaluates its right-hand side if and
+only if its left-hand side evaluates to nonzero.)
+
+$(P $(GLINK ConditionalExpression) evaluates its left-hand side argument
+first. Then, if the result is nonzero, the second operand is evaluated. Otherwise, the third operand
+is evaluated.)
+
+$(P Calls to functions  with `extern(D)` $(DDSUBLINK spec/attribute, linkage, linkage) (which is
+the default linkage) are evaluated in the following order: first, if necessary, the address of the
+function to call is evaluated (e.g. in the case of a computed function pointer or delegate). Then,
+arguments are evaluated left to right. Finally, transfer is passed to the function. Example:)
+
+$(SPEC_RUNNABLE_EXAMPLE_COMPILE
+---
+import std.stdio;
+void function(int a, int b, int c) fun()
+{
+    writeln("fun() called");
+    static void r(int a, int b, int c) { writeln("callee called"); }
+    return &r;
+}
+int g() { writeln("g() called"); return 1; }
+int h(int x) { writeln("h() called"); return x + 1; }
+int i() { writeln("i() called"); return 2; }
+int j() { writeln("j() called"); return 3; }
+fun()(g(), h(j()), i()); // evaluates fun() then g() then j() then h() then i()
+                         // after which control is transferred to the callee
```

Comment:
swap h and j's names, so you can write the order of evaluation in alphabetical order, which makes it easier to follow.

Reply:
Fixed